File: hphp/hack/src/typing/typing_per_cont_env.ml
(*
 * Copyright (c) 2015, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the "hack" directory of this source tree.
 *
 *)
open Common
module C = Typing_continuations
module CMap = C.Map

type per_cont_entry = {
  (* Local types per continuation. For example, the local types of the
   * break continuation correspond to the local types that there were at the
   * last encountered break in the current scope. These are kept to be merged
   * at the appropriate merge points. *)
  local_types: Typing_local_types.t;
  (* Fake members are used when we want member variables to be treated like
   * locals. We want to handle the following:
   * if($this->x) {
   *   ... $this->x ...
   * }
   * The trick consists in replacing $this->x with a "fake" local. So that
   * all the logic that normally applies to locals is applied in cases like
   * this. Hence the name: FakeMembers.
   * All the fake members are thrown away at the first call.
   * We keep the invalidated fake members for better error messages.
   *)
  fake_members: Typing_fake_members.t;
  (* Type parameter environment
   * Lower and upper bounds on generic type parameters and abstract types
   * For constraints of the form Tu <: Tv where both Tu and Tv are type
   * parameters, we store an upper bound for Tu and a lower bound for Tv.
   * Contrasting with tenv and subst, bounds are *assumptions* for type
   * inference, not conclusions.
   *)
  tpenv: Type_parameter_env.t;
}

type t = per_cont_entry C.Map.t

(*****************************************************************************)
(* Functions dealing with continuation based flow typing of local variables *)
(*****************************************************************************)

let empty_entry =
  {
    local_types = Typing_local_types.empty;
    fake_members = Typing_fake_members.empty;
    tpenv = Type_parameter_env.empty;
  }

let initial_locals entry = CMap.add C.Next entry CMap.empty

let get_cont_option = CMap.find_opt

let all_continuations : t -> C.t list = C.Map.keys

(** Continuations used to typecheck the `finally` block. *)
let continuations_for_finally =
  [C.Break; C.Continue; C.Catch; C.Exit; C.Finally]

(* Update an entry if it exists *)
let update_cont_entry name m f =
  match CMap.find_opt name m with
  | None -> m
  | Some entry -> CMap.add name (f entry) m

let add_to_cont name key value m =
  match CMap.find_opt name m with
  | None -> m
  | Some cont ->
    let cont =
      { cont with local_types = Local_id.Map.add key value cont.local_types }
    in
    CMap.add name cont m

let remove_from_cont name key m =
  match CMap.find_opt name m with
  | None -> m
  | Some c ->
    CMap.add
      name
      { c with local_types = Local_id.Map.remove key c.local_types }
      m

let drop_cont = CMap.remove

let drop_conts conts map =
  List.fold ~f:(fun map cont -> drop_cont cont map) ~init:map conts

let replace_cont key valueopt map =
  match valueopt with
  | None -> drop_cont key map
  | Some value -> CMap.add key value map

File: hphp/hack/src/typing/typing_per_cont_env.rs
// Copyright (c) Facebook, Inc. and its affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the "hack" directory of this source tree.

use arena_trait::Arena;
use typing_collections_rust::Map;

pub use crate::typing_continuations::TypingContKey;
use crate::typing_env_types::LocalId;
use crate::typing_local_types::{Local, TypingLocalTypes};

// TODO(hrust): Should this really be copy? Or should we keep a map of pointers.
#[derive(Debug, Clone, Copy)]
pub struct PerContEntry<'a> {
    pub local_types: TypingLocalTypes<'a>,
    // TODO(hrust): fake_members
    // TODO(hrust): tpenv
}
impl arena_trait::TrivialDrop for PerContEntry<'_> {}

impl<'a> PerContEntry<'a> {
    pub fn to_oxidized<'b>(
        &self,
        arena: &'b impl Arena,
    ) -> oxidized_by_ref::typing_per_cont_env::PerContEntry<'b>
    where
        'a: 'b,
    {
        let PerContEntry { local_types } = self;
        oxidized_by_ref::typing_per_cont_env::PerContEntry {
            local_types: arena_collections::map::Map::from(
                arena,
                local_types.iter().map(|(&k, &v)| (k, v.to_oxidized())),
            ),
            fake_members: oxidized_by_ref::typing_fake_members::TypingFakeMembers {
                valid: Default::default(),
                invalid: Default::default(),
            },
            tpenv: oxidized_by_ref::type_parameter_env::TypeParameterEnv {
                tparams: Default::default(),
                consistent: true,
            },
        }
    }
}

impl<'a> PerContEntry<'a> {
    pub fn empty() -> Self {
        PerContEntry {
            local_types: TypingLocalTypes::empty(),
        }
    }

    pub fn with_local_types<A: Arena>(
        &self,
        arena: &'a A,
        local_types: TypingLocalTypes<'a>,
    ) -> &'a Self {
        let mut entry = self.clone();
        entry.local_types = local_types;
        arena.alloc(entry)
    }
}

#[derive(Debug, Copy, Clone)]
pub struct TypingPerContEnv<'a>(Map<'a, TypingContKey<'a>, &'a PerContEntry<'a>>);

impl<'a> TypingPerContEnv<'a> {
    pub fn to_oxidized<'b>(
        &self,
        arena: &'b impl Arena,
    ) -> oxidized_by_ref::typing_per_cont_env::TypingPerContEnv<'b>
    where
        'a: 'b,
    {
        let TypingPerContEnv(map) = self;
        arena_collections::map::Map::from(
            arena,
            map.iter().map(|(&k, &v)| (k, v.to_oxidized(arena))),
        )
    }
}

impl<'a> TypingPerContEnv<'a> {
    pub fn initial_locals<A: Arena>(arena: &'a A, entry: PerContEntry<'a>) -> Self {
        let m = Map::empty();
        let entry: &'a PerContEntry<'a> = arena.alloc(entry);
        let m = m.add(arena, TypingContKey::Next, entry);
        TypingPerContEnv(m)
    }

    pub fn get_cont_option(self, name: TypingContKey<'a>) -> Option<&'a PerContEntry<'a>> {
        self.0.find(&name)
    }

    pub fn add_to_cont<A: Arena>(
        self,
        arena: &'a A,
        name: TypingContKey<'a>,
        key: LocalId<'a>,
        value: Local<'a>,
    ) -> Self {
        let m = self.0;
        let m = match m.find(&name) {
            None => m,
            Some(cont) => {
                let value = arena.alloc(value);
                let local_types = cont.local_types.add(arena, key, value);
                let cont = cont.with_local_types(arena, local_types);
                m.add(arena, name, cont)
            }
        };
        TypingPerContEnv(m)
    }
}
